### PR TITLE
feat(chronorelais): add endpoint relay points list

### PR DESCRIPTION
**Pull request by THOMASPH2M**
### Why
We need to get list of relay points from a specific address.

### How
We use existing code from Chronorelais Magento module.

Params example required (same as Chronorelais module)

`{
  "postCode":"38290", // required
  "shippingAddress":{
    "country_id":"FR", // required
    "street":"714 route de la ruette",
    "postcode":"38290",
    "city":"Satolas-Et-Bonce"
  }
}`

### CODE REVIEW

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            throw new Exception(__('Missing shipping address postcode'));
+        }
+
+        if (!isset($shippingAddress['street'])) {
+            throw new Exception(__('Missing shipping address street'));
+        }
+
+        if (!isset($shippingAddress['city'])) {
+            throw new Exception(__('Missing shipping address postcode'));
+        }
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php]: those are not mandatory and the API should be smart to work without them.

**Review by dpobel — COMMENTED** [review]
besides the remark, that looks ok

**Comment by THOMASPH2M** [comment]
PR  updated :heavy_check_mark:

File: app/code/FrontCommerce/ChronorelaisHeadless/composer.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/composer.json:1]: to be installable using composer, a composer.json is needed at the root of the  repository. I copy pasted this one but if I understand correctly those 2 composer.json probably need to be a bit different.

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    /**
+     * @throws Exception
+     */
+    protected function checkParams(string $postCode, array $shippingAddress)
+    {
+        if (!$postCode) {
+            throw new Exception(__('Missing post_code'));
+        }
+
+        if (!isset($shippingAddress['country_id'])) {
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php]: this does not follow what is written in the ticket.

The API is supposed to respond to `GET /rest/default/V1/chronorelais/relais?country_id=<COUNTRYCODE>&postcode=<ZIPCODE>` with that it expects `GET /rest/default/V1/chronorelais/relais?shippingAddress[country_id]=<COUNTRYCODE>&postcode=<ZIPCODE>`

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+        return $result;
+    }
+
+    /**
+     * @throws Exception
+     */
+    protected function checkParams(string $postCode, array $shippingAddress)
+    {
+        if (!$postCode) {
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php]: here as well, there's an issue. The parameter is supposed to be in lower case.

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPoint.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        $shippingAddress = $this->addressFactory->create()->setData($shippingAddress);
+
+        return $this->helperWebservice->getPointRelaisByAddress(
+            $methodCode,
+            $shippingAddress
+                ->setData('postcode', $postCode)
+                ->setData('city', $shippingAddress['city'] ?: 'unknown')
+                ->setData('country', $shippingAddress['country_id'] ?: 'unknown')
+                ->setData('street', $shippingAddress['street'] ?: 'unknown')
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPoint.php]: and there as well, the post code parameter is supposed to be in lowercase, the other parameters are not supposed to be called `shippingAddress[whatever]` but directly `whatever`.

:warning: `street` is supposed to be array because the customer address might have several lines.

In a nutshell, please make sure the URL in the ticket work ie:

```
# minimum parameters
GET /rest/default/V1/chronorelais/relais?country_id=<COUNTRYCODE>&postcode=<ZIPCODE>
# should work as well
GET /rest/default/V1/chronorelais/relais?country_id=<COUNTRYCODE>&postcode=<ZIPCODE>&city=<CITY>
# complete parameter list
GET /rest/default/V1/chronorelais/relais?country_id=<COUNTRYCODE>&street[]=<STREETVALUE>&postcode=<ZIPCODE>&city=<CITY>
# with several line in the street
GET /rest/default/V1/chronorelais/relais?country_id=<COUNTRYCODE>&street[]=<STREETVALUE1>&street[]=<STREETVALUE2>&postcode=<ZIPCODE>&city=<CITY>
```

File: app/code/FrontCommerce/ChronorelaisHeadless/etc/webapi.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<routes xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Webapi:etc/webapi.xsd">
+    <route method="GET" url="/V1/chronorelais/relay-point">
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/etc/webapi.xml]: should be:
```suggestion
    <route method="GET" url="/V1/chronorelais/relais">
```

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+                ->setName($relayPoint->nomEnseigne)
+                ->setOpeningTime([
+                    'Monday' => $relayPoint->horairesOuvertureLundi,
+                    'Tuesday' => $relayPoint->horairesOuvertureMardi,
+                    'Wednesday' => $relayPoint->horairesOuvertureMercredi,
+                    'Thursday' => $relayPoint->horairesOuvertureJeudi,
+                    'Friday' => $relayPoint->horairesOuvertureVendredi,
+                    'Saturday' => $relayPoint->horairesOuvertureSamedi,
+                    'Sunday' => $relayPoint->horairesOuvertureDimanche,
+                ]);
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php]: this results in:

```json
{
  "relay_id": "6055S",
  "street1": "6 rue Saint Louis",
  "street2": "",
  "street3": "",
  "latitude": "43.61611111110",
  "longitude": "3.15583333333",
  "zip_code": "34600",
  "city": "BEDARIEUX",
  "name": "TABAC LOTO DU PONT D ORB",
  "opening_time": [
    "07:00-12:00 12:00-19:30",
    "07:00-12:00 12:00-19:00",
    "07:00-12:00 12:00-19:00",
    "07:00-12:00 12:00-19:00",
    "07:00-12:00 12:00-19:00",
    "07:00-12:00 12:00-19:00",
    "00:00-00:00 00:00-00:00"
  ]
}
```

while the expected format is:

```json
{
    "street1": "64 Rue principale",
    "street2": "",
    "street3": "",
    "latitude": "46.13221716813",
    "longitude": "5.16451953636",
    "zipcode": "01960",
    "id": "2039V",
    "city": "Servas",
    "name": "Tabac au rendez vous des amis",
    "openingTime": {
        "monday": "07:00-12:00 12:00-18:30",
        "tuesday": "07:00-12:00 12:00-18:30",
        "wednesday": "07:00-12:00 12:00-18:30",
        "thursday": "07:00-12:00 12:00-18:30",
        "friday": "07:00-12:00 12:00-18:30",
        "saturday": "07:00-12:00 12:00-18:30",
        "sunday": "07:00-12:00 12:00-13:00"
    }
}
```

the camelcase vs. snakecase and the id vs. relay_id changes are not a big issue but not having the days in opening time is an issue.

**Review by dpobel — CHANGES_REQUESTED** [review]

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php (hunk 7 above)
Comment (dpobel) [reply app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php]: openingTime is still not correct:

```json
{
    "street1": "6 rue Saint Louis",
    "street2": "",
    "street3": "",
    "latitude": "43.61611111110",
    "longitude": "3.15583333333",
    "zipcode": "34600",
    "id": "6055S",
    "city": "BEDARIEUX",
    "name": "TABAC LOTO DU PONT D ORB",
    "openingTime": [
      "07:00-12:00 12:00-19:30",
      "07:00-12:00 12:00-19:00",
      "07:00-12:00 12:00-19:00",
      "07:00-12:00 12:00-19:00",
      "07:00-12:00 12:00-19:00",
      "07:00-12:00 12:00-19:00",
      "00:00-00:00 00:00-00:00"
    ]
  }
```

the days are missing

File: app/code/FrontCommerce/ChronorelaisHeadless/Api/RelayPointManagementInterface.php [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+namespace FrontCommerce\ChronorelaisHeadless\Api;
+
+interface RelayPointManagementInterface
+{
+    /**
+     * @param string $postcode
+     * @param string $countryId
+     * @param string $city
+     * @param string[] $street
+     * @return string[]
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Api/RelayPointManagementInterface.php]: an array of strings ?

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPoint.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+     */
+    public function getRelayPoints(
+        string $methodCode,
+        string $postcode,
+        string $countryId,
+        string $city,
+        array  $street
+    ): array
+    {
+        $street = $street ? implode(' ', $street) : '';
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPoint.php]: unless I'm mistaken, it can be simplified with:

```suggestion
        $street = implode(' ', $street);
```

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    protected function formatRelayResponseData(array $relayPoints): array
+    {
+        if (!$relayPoints) {
+            return [];
+        }
+
+        $result = [];
+        $relayPoints = json_decode(json_encode($relayPoints));
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php]: why is this needed ?

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+        foreach ($relayPoints as $relayPoint) {
+            $relayPoint = json_decode(json_encode($relayPoint));
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php]: same question here

File: app/code/FrontCommerce/ChronorelaisHeadless/Plugin/ShippingInformationManagement.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    protected CheckoutSession $checkoutSession;
+    protected CartRepositoryInterface $quoteRepository;
+    protected HelperWebservice $helperWebservice;
+    protected Data $helper;
+
+    public function __construct(
+        CheckoutSession         $checkoutSession,
+        CartRepositoryInterface $cartRepository,
+        HelperWebservice        $webservice,
+        Data                    $helper
```
Comment (dpobel) [inline app/code/FrontCommerce/ChronorelaisHeadless/Plugin/ShippingInformationManagement.php:24]: those do not seem to be used, so they can be removed, right ?

**Review by dpobel — CHANGES_REQUESTED** [review]

File: app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php [hunk 13]
```diff
@@ -54,45 +48,31 @@ protected function formatRelayResponseData(array $relayPoints): array
...
-                ->setOpeningTime([
-                    'Monday' => $relayPoint->horairesOuvertureLundi,
-                    'Tuesday' => $relayPoint->horairesOuvertureMardi,
-                    'Wednesday' => $relayPoint->horairesOuvertureMercredi,
-                    'Thursday' => $relayPoint->horairesOuvertureJeudi,
-                    'Friday' => $relayPoint->horairesOuvertureVendredi,
-                    'Saturday' => $relayPoint->horairesOuvertureSamedi,
-                    'Sunday' => $relayPoint->horairesOuvertureDimanche,
-                ]);
+            $result[] = [
```
Comment (THOMASPH2M) [inline app/code/FrontCommerce/ChronorelaisHeadless/Model/RelayPointManagement.php:47]: Le changement de la response @dpobel
